tests/repository/test_repo_lock.py

```diff
@@ -0,0 +1,164 @@
+        datetime.fromtimestamp(1120.0),
+        datetime.fromtimestamp(1120.0),
+        datetime.fromtimestamp(1120.0),
+        datetime.fromtimestamp(1180.0),
+    ]
+    random_mock.choices = Mock(return_value="1234")
+    # Pulp API Calls
+    # 1. Get the current locks to check if three's stale ones. None this time.
+    # 2. Put our lock in the repo notes.
+    # 3. Get the current locks to check if it's our turn to run. (not yet)
+    # 3. Get the current locks again, it's our turn this time.
+    # 4. Delete the lock from the repo notes.
+    mock_client._get_repo_lock_data.side_effect = [
+        f_return({}),
+        f_return(
+            {
+                "pulplib-lock-claim-1233": '{"id": "1233", "context": "testing repo lock", "created": 900.0, "valid_from": 910.0, "expires": 1900.0}',
+                "pulplib-lock-claim-1234": '{"id": "1234", "context": "testing repo lock", "created": 1000.0, "valid_from": 1010.0, "expires": 2000.0}',
+            }
+        ),
+        f_return(
+            {
+                "pulplib-lock-claim-1234": '{"id": "1234", "context": "testing repo lock", "created": 1000.0, "valid_from": 1010.0, "expires": 2000.0}'
+            }
+        ),
+    ]
+    expected_mock_calls = [
+        call(
+            "test-repo",
+            {
+                "pulplib-lock-claim-1234": '{"id": "1234", "context": "testing repo lock", "created": 1000.0, "valid_from": 1010.0, "expires": 2000.0}'
+            },
+        ),
+        call("test-repo", {"pulplib-lock-claim-1234": None}, await_result=True),
+    ]
+    lock = RepoLock(
+        "test-repo",
+        mock_client,
+        "testing repo lock",
+        1000,
+    )
+
+    # Runs __enter__ and __exit__ in RepoLock
+    with lock:
+        pass
+
+    mock_client._update_repo_lock_data.assert_has_calls(expected_mock_calls)
+
+    assert mock_client._get_repo_lock_data.call_count == 3
+
+
+@patch("pubtools.pulplib._impl.model.repository.repo_lock.now")
+@patch("pubtools.pulplib._impl.model.repository.repo_lock.random")
+def test_repo_lock_fake_client(random_mock, datetime_mock):
+    datetime_mock.side_effect = [
+        datetime.fromtimestamp(1000.0),
+        datetime.fromtimestamp(1060.0),
+        datetime.fromtimestamp(1120.0),
+        datetime.fromtimestamp(1120.0),
+        datetime.fromtimestamp(1120.0),
+        datetime.fromtimestamp(1180.0),
+    ]
+    random_mock.choices = Mock(return_value="1234")
+    ctrl = FakeController()
+    ctrl.insert_repository(Repository(id="test-repo"))
+    lock = ctrl.new_client().get_repository("test-repo").lock("test")
+    expected_calls = [
+        {
+            "pulplib-lock-claim-1234": '{"id": "1234", "context": "test", "created": 1000.0, "valid_from": 1010.0, "expires": 2800.0}'
+        },
+        {"pulplib-lock-claim-1234": None},
+    ]
+
+    # Runs __enter__ and __exit__ in RepoLock
+    with lock:
+        pass
+
+    assert ctrl._state.repo_lock_history == expected_calls
```

Comment:
A couple of issues with this:

- the `_state` property isn't public API, so there still isn't a proper way to use this from other projects.  Properties on `_state` which are meant to be public can be accompanied by an accessor on FakeController class, which is also where you'd write a doc string for it.
- the format is nasty - string values.  Is the user supposed to parse them as JSON?  We should avoid exposing the implementation detail that locks are achieved by updating repo notes which happens to use JSON documents.

Ideally this would be consistent with publish_history, sync_history and return named tuples.